subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java

```diff
@@ -691,45 +692,37 @@ private boolean doesConsumerDependOnDestroyer(Node consumer, Node destroyer) {
     }
 
     private static boolean hasOverlap(Iterable<String> paths1, Iterable<String> paths2) {
-        for (String path1 : paths1) {
-            for (String path2 : paths2) {
-                String overLappedPath = getOverLappedPath(path1, path2);
-                if (overLappedPath != null) {
+        //build a helper TreeSet with the contents of paths2
+        final TreeSet<String> set = new TreeSet<>();
+        for (final String path : paths2) {
+            set.add(path);
+        }
+
+        for (final String path1 : paths1) {
+            //check the path verbatim
+            if (set.contains(path1)) {
+                return true;
+            }
+            //first iterate over all path segments (prefixes)
+            int endIndex = path1.indexOf(File.separatorChar);
+            while (endIndex >= 0) {
+                final String currentPrefix = path1.substring(0, endIndex);
+                if (set.contains(currentPrefix)) {
                     return true;
                 }
+                endIndex = path1.indexOf(File.separatorChar, endIndex + 1);
+            }
```

Comment:
Is this iteration required?

Reply:
I think it is, otherwise no match would be found for "a/b/c" in the map {"a/b"}, for example. Also, the iteration is probably not that bad performance-wise because it basically amounts to going through the string once (which the original code also did implicitly in the .startsWith method)

Reply:
You are calling `substring(0, endIndex)` multiple times, which has overall quadratic complexity with regards to the number of `/` in the path.
As you said, the paths might be long.

Reply:
Can `a/b/c` be found with something like map.ceiling / floor?

Reply:
I don't think so, at least I had no idea how it could be done fast and easy. As an example, take "a/c/d" that shall be looked up in the map {"a/b/c", "a/c", "a/c/a"}. Both floor and ceiling will not lead to the right entry. I could iterate, but I doubt that that's more efficient.

That said, your remark sparked another idea for an optimization. I guess that the paths usually all have the same prefix (i.e., the common project directory). In this case, this common prefix can be skipped. I added another commit with this optimization. And I did a test with more realistic string lengths. Results:
10x10: Old algorithm: ~0ms, new algorithm: ~0ms
10000x10: Old algorithm: 1.1ms, new algorithm: 0.7ms
10x10000: Old algorithm: 1.9ms, new algorithm: 3.4ms
10000x10000: Old algorithm: 2010.0ms, new algorithm: 6.4ms